src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -140,264 +35,133 @@ EFeatureType RandomFourierGaussPreproc::get_feature_type() {
-	}
+	SGVector<float64_t> res(m_dims);
+	SGVector<float64_t> offset_vector(m_dims);
 
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
+	if(m_basis.size() == 0 || m_offset.size() == 0)
+		error("Random coefficients not instantiated, use fit() before");
 
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
+	linalg::add(m_offset,linalg::matrix_prod(m_basis,vector),offset_vector);
+
+	for(auto idx : range(m_dims))
+		res[idx] = std::sqrt(2.0 / m_dims) * cos(offset_vector[idx]);
 
 	return res;
 }
 
-SGMatrix<float64_t>
-RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
+SGMatrix<float64_t> RandomFourierGaussPreproc::apply_to_matrix(const SGMatrix<float64_t>& matrix)
 {
-	// version for case dim_feature_space < dim_input space with direct transformation on feature matrix ??
-
 	auto num_vectors = matrix.num_cols;
 	auto num_features = matrix.num_rows;
 
-	io::info("get Feature matrix: {}x{}", num_vectors, num_features);
+	/* Check if random coefficients are instantiated */
+	if(m_basis.size() == 0 || m_offset.size() == 0)
+		error("Random coefficients not instantiated, use fit() before");
 
-	if (num_features != cur_dim_input_space)
-	{
-		throw ShogunException(
-		    "float64_t * "
-		    "CRandomFourierGaussPreproc::apply_to_matrix("
-		    "SGMatrix<float64_t> matrix): matrix.num_rows != "
-		    "cur_dim_input_space is not allowed\n");
-	}
+	io::info("get Feature matrix: {}x{}", num_vectors, num_features);
 
-	SGMatrix<float64_t> res(cur_dim_feature_space, num_vectors);
+	SGMatrix<float64_t> res(m_dims, num_vectors);
+	
+	/* Temporary matrix to store offsetted matrix product*/
+	SGMatrix<float64_t> offset_mat(m_dims,num_vectors);
 
-	auto val = std::sqrt(2.0 / cur_dim_feature_space);
+	linalg::add_vector(linalg::matrix_prod(m_basis,matrix),
+					   m_offset,
+					   offset_mat);
+	res = std::sqrt(2.0/m_dims) * cos(offset_mat);
 
-	for (auto vec : range(num_vectors))
-	{
-		for (auto od : range(cur_dim_feature_space))
-		{
-			SGVector<float64_t> a(
-			    matrix.matrix + vec * num_features, cur_dim_input_space, false);
-			SGVector<float64_t> b(
-			    randomcoeff_multiplicative + od * cur_dim_input_space,
-			    cur_dim_input_space, false);
-			res(od, vec) =
-			    val * cos(randomcoeff_additive[od] + linalg::dot(a, b));
-		}
-	}
-
-	return res;
+	return res;	
 }
 
 void RandomFourierGaussPreproc::cleanup()
 {
-
+	m_basis = SGMatrix<float64_t>();
```

Comment:
no need for this method

Reply:
This is defined purely virtual from the class it's inheriting. If we don't define it here, this class will become abstract.

Reply:
Then you can make it empty for now (and later we will remove it in the base class)

Reply:
Second thoughts, wouldn't it be nice to keep it?. It's a handy method to reset internal members. In our fit method, we could do 
```c++
void RFGP::fit(Features f)
if(m_fitted)
  cleanup();
....
init_basis(f);
m_fitted = true;
```
It all falls nicely together, and as you said: "we don't delete stuff" ;)

Reply:
no need to clean up. Applying fit again simply overwrites previous states